src/agentscope/service/web_search/web_digest.py

```diff
@@ -0,0 +1,273 @@
+        "Cache-Control": "max-age=0",
+        "Upgrade-Insecure-Requests": "1",
+        "User-Agent": "Mozilla/5.0 (X11; Linux x86_64)"
+        " AppleWebKit/537.36 (KHTML, like Gecko) ",
+    }
+    try:
+        response = requests.get(url=url, headers=header, timeout=timeout)
+        if response.status_code == 200:
+            content_type = response.headers["Content-Type"].lower()
+            if "html" in content_type:
+                return parse_html(
+                    response.text,
+                    keep_raw,
+                    html_selected_tags,
+                    html_parse_func,
+                )
+            elif "pdf" in content_type:
+                # TODO: support pdf in the future
+                raise NotImplementedError(
+                    "Unsupported url to PDF content.",
+                )
+            elif "json" in content_type:
+                return ServiceResponse(
+                    ServiceExecStatus.SUCCESS,
+                    content=json.loads(response.text),
+                )
+            elif "image" in content_type:
+                # TODO: to support image (gif, jpeg, png) data
+                logger.warning(
+                    "Current implementation returns binary "
+                    "response.content for url with image Content-Types",
+                )
+                return ServiceResponse(
+                    ServiceExecStatus.SUCCESS,
+                    content=response.content,
+                )
+            else:
+                raise NotImplementedError(
+                    f"Unsupported content type ({content_type}) "
+                    f"with url: ({url})",
+                )
+        else:
+            logger.warning(
+                f"Fail to load web page, "
+                f"status code {response.status_code}",
+            )
+            return ServiceResponse(
+                ServiceExecStatus.ERROR,
+                content="",
+            )
+    except Exception as e:
+        logger.warning(e)
+        return ServiceResponse(ServiceExecStatus.ERROR, content="")
+
+
+def parse_html(
+    html_text: str,
+    keep_raw: bool = True,
+    html_selected_tags: Optional[Sequence[str]] = None,
+    html_parse_func: Optional[Callable] = None,
+) -> ServiceResponse:
+    """
+    Parse the obtained HTML file.
+    Args:
+        html_text (str):
+            HTTP response text
+        keep_raw (bool):
+            Whether to keep raw HTML. If True, the content is
+            stored with key "raw".
+        html_selected_tags (Optional[Sequence[str]]):
+            the text in elements of `selected_tags` will
+            be extracted and stored with "selected_tags_text"
+            key in return.
+        html_parse_func (Optional[Callable]):
+            if "html_parse_func" is not None, then the
+            `html_parse_func` will be invoked with the response
+            text as input. The result is stored with
+            `self_define_func` key
```

Comment:
There is no `self_define_func` key in lines 172-177.

Reply:
fixed